python/ray/serve/batching.py

```diff
@@ -0,0 +1,275 @@
+            # queue.
+            if curr_timeout == 0:
+                batch.append(await self.queue.get())
+            # If the timeout is nonzero, wait for either the timeout to occur
+            # or the max batch size to be ready.
+            else:
+                try:
+                    await asyncio.wait_for(self.full_batch_event.wait(),
+                                           curr_timeout)
+                except asyncio.TimeoutError:
+                    pass
+
+            # Pull up to the max_batch_size requests off the queue.
+            while len(batch) < self.max_batch_size and not self.queue.empty():
+                batch.append(self.queue.get_nowait())
+
+            # Reset the event if there are fewer than max_batch_size requests
+            # in the queue.
+            if (self.queue.qsize() < self.max_batch_size
+                    and self.full_batch_event.is_set()):
+                self.full_batch_event.clear()
+
+            # Adjust the timeout based on the time spent in this iteration.
+            curr_timeout = max(0, curr_timeout - (time.time() - loop_start))
+
+        return batch
+
+    async def _handle_batches(self, func):
+        while True:
+            batch = await self.wait_for_batch()
+            assert len(batch) > 0
+            self_arg = batch[0][0]
+            args = [item[1] for item in batch]
+            futures = [item[2] for item in batch]
+
+            try:
+                # Method call.
+                if self_arg is not None:
+                    results = await func(self_arg, args)
+                # Normal function call.
+                else:
+                    results = await func(args)
+
+                if len(results) != len(batch):
+                    raise RayServeException(
+                        "Batched function doesn't preserve batch size. "
+                        f"The input list has length {len(batch)} but the "
+                        f"returned list has length {len(results)}.")
+            except Exception as e:
+                results = [_RaisedExceptionWrapper(e)] * len(batch)
+
+            for i, result in enumerate(results):
+                futures[i].set_result(result)
+
+    def __del__(self):
+        if self._handle_batch_task is None:
+            return
+
+        async def await_task():
+            try:
+                await self._handle_batch_task
+            except asyncio.CancelledError:
+                pass
+
+        # TODO(edoakes): although we try to gracefully shutdown here, it still
+        # causes some errors when the process exits due to the asyncio loop
+        # already being destroyed.
+        self._handle_batch_task.cancel()
+        asyncio.get_event_loop.run_until_complete(await_task())
+
+
+T = TypeVar("T")
+R = TypeVar("R")
+F = TypeVar("F", bound=Callable[[List[T]], List[R]])
+G = TypeVar("G", bound=Callable[[T], R])
+
+
+# Normal decorator use case (called with no arguments).
```

Comment:
ah this is probably good enough. when we have fully typed check Ray we can revisit this